[mlir] MemRefToSPIRV propagate alignment attributes from MemRef ops.

amd-eochoalo opened the pull request:
This patchset:
* propagates alignment attributes from memref operations into the SPIR-V dialect,
* fixes an error in the logic which previously propagated alignment attributes but did not add other MemoryAccess attributes.
* adds a failure condition in the case where the alignment attribute from the memref dialect (64-bit wide) does not fit in SPIR-V's alignment attribute (specified to be 32-bit wide).

amd-eochoalo committed:
[mlir] Use memref's alignment attribute directly.

amd-eochoalo committed:
[mlir] MemRefToSPIRV propagate alignment attribute.

kuhar commented:
> * It is unclear to me at the moment whether non-PhysicalStorageBuffers may have an alignment attribute. At the moment I am assuming they won't.

PSB requires alignment, and in the other storage classes alignment is optional.

> * The alignment attribute added to the SPIR-V operations is 32-bits wide while the alignment attribute in MemRef operations is 64-bits wide. I would recommend changing this to 64-bits [per this comment. ](https://github.com/llvm/llvm-project/pull/144344#discussion_r2150375562). However, the current definition constraints alignment to be 32 bits wide.

This is because the SPIR-V spec specifies alignment as a 32-bit power-of-two number.

amd-eochoalo committed:
[mlir] Fix calculateMemoryRequirements in MemRefToSPIRV.

There was an early return in calculateMemoryRequirements
that looked explicitly for alignment and only set the alignment attribute.
However, this was not correct for the following reasons:

* Alignment was set only if both the alignment and the
  memory_access attributes were both present in the memref operation,
  without handling the case when only the alignment was exclusively
  present.
* In the case alignment and memory_access attributes were both present,
  the memory_access attribute would not be updated to aligned if
  the memory_access attribute was not marked aligned.
* In the case alignment and memory_access attributes were both present,
  other memory requirements (e.g., non_temporal) would not be added
  as attributes.

amd-eochoalo committed:
[mlir] Ensure memref's alignment is within SPIRV's alignment bounds.

amd-eochoalo committed:
[mlir] Add aligned nontemporal load test to MemRefToSPIRV.

amd-eochoalo commented on mlir/test/Conversion/MemRefToSPIRV/memref-to-spirv.mlir:
```diff
@@ -86,8 +86,23 @@ func.func @load_i1(%src: memref<4xi1, #spirv.storage_class<StorageBuffer>>, %i :
+  // CHECK: %[[VAL:.+]] = spirv.Load "StorageBuffer" %[[ADDR]] ["Aligned", 32] : i8
+  // CHECK: %[[ZERO_I8:.+]] = spirv.Constant 0 : i8
+  // CHECK: %[[BOOL:.+]] = spirv.INotEqual %[[VAL]], %[[ZERO_I8]] : i8
+  %0 = memref.load %src[%i] { alignment = 32 } : memref<4xi1, #spirv.storage_class<StorageBuffer>>
```
Thanks for the suggestion! I've added it here: https://github.com/llvm/llvm-project/pull/151723/commits/0337e226ce41bbcc4b57c2e70186c510361afe80

amd-eochoalo committed:
Address review comments
